ot/mpz-ot-core/src/kos/receiver.rs

```diff
@@ -0,0 +1,486 @@
+        let chis = (0..unchecked_ts.len())
+            .map(|_| Block::random(&mut rng))
+            .collect::<Vec<_>>();
+
+        // Figure 7, "Check correlation", point 2.
+        // Compute the random linear combinations.
+        cfg_if::cfg_if! {
+            if #[cfg(feature = "rayon")] {
+                let (x, t0, t1) = unchecked_choices.par_iter()
+                    .zip(&unchecked_ts)
+                    .zip(chis)
+                    .map(|((c, t), chi)| {
+                        let x = if *c { chi } else { Block::ZERO };
+                        let (t0, t1) = t.clmul(chi);
+                        (x, t0, t1)
+                    })
+                    .reduce(
+                        || (Block::ZERO, Block::ZERO, Block::ZERO),
+                        |(_x, _t0, _t1), (x, t0, t1)| {
+                            (_x ^ x, _t0 ^ t0, _t1 ^ t1)
+                        },
+                    );
+            } else {
+                use itybity::ToBits;
+
+                let (x, t0, t1) = unchecked_choices.iter()
+                    .zip(&unchecked_ts)
+                    .zip(chis)
+                    .map(|((c, t), chi)| {
+                        let x = if *c { chi } else { Block::ZERO };
+                        let (t0, t1) = t.clmul(chi);
+                        (x, t0, t1)
+                    })
+                    .reduce(|(_x, _t0, _t1), (x, t0, t1)| {
+                        (_x ^ x, _t0 ^ t0, _t1 ^ t1)
+                    }).unwrap();
+            }
+        }
+
+        // Strip off the rows sacrificed for the consistency check.
+        let nrows = unchecked_ts.len() - (CSP + SSP);
+        unchecked_ts.truncate(nrows);
+        unchecked_choices.truncate(nrows);
+
+        cfg_if::cfg_if! {
+            if #[cfg(feature = "rayon")] {
+                let iter = unchecked_ts.par_iter().enumerate();
+            } else {
+                let iter = unchecked_ts.iter().enumerate();
+            }
+        }
+
+        let cipher = &(*FIXED_KEY_AES);
+        let keys = iter
+            .map(|(j, t)| {
+                let j = Block::from(((self.state.key_counter + j) as u128).to_be_bytes());
+                cipher.tccr(j, *t)
+            })
+            .collect::<Vec<_>>();
+
+        self.state.key_counter += keys.len();
+
+        // Add to existing keys.
+        self.state.keys.extend(keys);
+        self.state.choices.extend(unchecked_choices);
+
+        // If we're recording, we track `ts` too
+        if self.tape.is_some() {
+            self.state.ts.extend(unchecked_ts);
+        }
+
+        Check { x, t0, t1 }
+    }
+
+    /// Derandomize the receiver's choices from the setup phase.
+    ///
+    /// # Arguments
+    ///
+    /// * `choices` - The receiver's corrected choices.
+    pub fn derandomize(&mut self, choices: &[bool]) -> Derandomize {
+        let flip = Vec::<u8>::from_lsb0_iter(
+            self.state
+                .choices
+                .iter()
```

Comment:
Shouldn't we return an error if `choices` > `self.state.choices`. I believe in this case it would silently drop the rest of `choices`.